### ISSUE TITLE
Bulkinsert not working Azure SQL DW

### ISSUE THREAD

**Issue by rajeshmese** [Azure/azure-sqldb-spark#43]
Writing the dataframe to AZURE SQL DW is working fine for normal jdbc write.sqlDB API.
But while using bulkCopyToSqlDB api getting data conversion error. I'm not providing bulkcopy metadata.

19/08/05 17:06:00 INFO DAGScheduler: Job 3 failed: foreachPartition at DataFrameFunctions.scala:72, took 26.016681 s
org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 5.0 failed 4 times, most recent failure: Lost task 0.3 in stage 5.0 (TID 11, wn3-hdi-gl.zeu1lfaty0eerhtaqktnvsba5f.ix.internal.cloudapp.net, executor 2): com.microsoft.sqlserver.jdbc.SQLServerException: An error occurred while converting the String value to JDBC data type DECIMAL.
	at com.microsoft.sqlserver.jdbc.SQLServerBulkCopy.writeColumnToTdsWriter(SQLServerBulkCopy.java:2528)
	at com.microsoft.sqlserver.jdbc.SQLServerBulkCopy.writeColumn(SQLServerBulkCopy.java:2766)
	at com.microsoft.sqlserver.jdbc.SQLServerBulkCopy.writeBatchData(SQLServerBulkCopy.java:3274)
	at com.microsoft.sqlserver.jdbc.SQLServerBulkCopy.doInsertBulk(SQLServerBulkCopy.java:1570)
	at com.microsoft.sqlserver.jdbc.SQLServerBulkCopy.access$200(SQLServerBulkCopy.java:58)
	at com.microsoft.sqlserver.jdbc.SQLServerBulkCopy$1InsertBulk.doExecute(SQLServerBulkCopy.java:709)
	at com.microsoft.sqlserver.jdbc.TDSCommand.execute(IOBuffer.java:7151)
	at com.microsoft.sqlserver.jdbc.SQLServerConnection.executeCommand(SQLServerConnection.java:2478)
	at com.microsoft.sqlserver.jdbc.SQLServerBulkCopy.sendBulkLoadBCP(SQLServerBulkCopy.java:739)
	at com.microsoft.sqlserver.jdbc.SQLServerBulkCopy.writeToServer(SQLServerBulkCopy.java:1684)
	at com.microsoft.sqlserver.jdbc.SQLServerBulkCopy.writeToServer(SQLServerBulkCopy.java:669)
	at com.microsoft.azure.sqldb.spark.connect.DataFrameFunctions.com$microsoft$azure$sqldb$spark$connect$DataFrameFunctions$$bulkCopy(DataFrameFunctions.scala:127)
	at com.microsoft.azure.sqldb.spark.connect.DataFrameFunctions$$anonfun$bulkCopyToSqlDB$1.apply(DataFrameFunctions.scala:72)
	at com.microsoft.azure.sqldb.spark.connect.DataFrameFunctions$$anonfun$bulkCopyToSqlDB$1.apply(DataFrameFunctions.scala:72)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:935)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:935)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:121)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:402)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:408)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassCastException: java.lang.String cannot be cast to java.math.BigDecimal
	at com.microsoft.sqlserver.jdbc.SQLServerBulkCopy.writeColumnToTdsWriter(SQLServerBulkCopy.java:2205)
	... 25 more

**Comment by david-nguyen** [comment]
There error indicates that the mapping is incorrect, double check the dataframe mapping to the table column.

**Comment by pkreddy** [comment]
@rajeshmese were you able to solve your issue ? I'm facing a similar issue but I'm writing to Azure SQL database.

**Comment by nijanthanvijayakumar** [comment]
@rajeshmese and @pkreddy I faced the same issue while writing from Spark to Azure SQL Server DB. 

There are somecases where you will get the same error as @rajeshmese or slightly a different error.

- Incorrect column order mapping.
- Incorrect data type mapping between DF columns and SQL Server table columns.
- Incorrect number of columns between DF and SQL Server table.

In this context, the error corresponds to the incorrect data type mapping of the DataFrame column and the SQL DB table column (check line #7). The SQL DB column is of DECIMAL type and the corresponding column in DataFrame is of STRING type. 

Use the right column data type mapping to solve such errors. I got an issue while writing a float value  (from DataFrame) to a corresponding column of double value (in SQL server DB). I made the conversion on the fly in my query by type casting the FLOAT to DOUBLE and fixed the issue. Hope this helps!
